drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
+                ROT_TO_SP_RESET_L_IN_PINT_SLOT,
+                PintOp::Clear,
+                PintCondition::Rising,
+            );
+            let _ = gpio.pint_op(
+                ROT_TO_SP_RESET_L_IN_PINT_SLOT,
+                PintOp::Clear,
+                PintCondition::Falling,
+            );
+        }
+    }
+
+    // The SP is halted at its reset vector.
+    fn sp_measure_fast(&mut self) -> Result<[u8; 256 / 8], ()> {
+        // write program entry address to PC (R15)
+        // write top-of-stack address to MSP (R13)
+        // write start of image/vector table address(0x20000000) to VTOR (at 0xe000ed08)
+        // write the program to RAM
+        // write DHCSR to RUN (MATIC + C_DEBUGEN)
+        // poll for S_HALT or timeout
+
+        // Set SP's Program Counter
+        let sp_reset_vector =
+            u32::from_le_bytes(ENDOSCOPE_BYTES[4..=7].try_into().unwrap_lite());
+        if self
+            .do_write_core_register(Reg::Dr.into(), sp_reset_vector | 1)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::WrotePcRegisterFail);
+            return Err(());
+        }
+
+        // Set SP's Stack Pointer
+        let sp_initial_sp =
+            u32::from_le_bytes(ENDOSCOPE_BYTES[0..=3].try_into().unwrap_lite());
+        if self
+            .do_write_core_register(Reg::Sp.into(), sp_initial_sp)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::WroteSpRegisterFail);
+            return Err(());
+        }
+
+        // Set VTOR - Set vector table base address
+        if self
+            .write_single_target_addr(VTOR, endoscope::LOAD)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::CannotWriteVtor);
+            return Err(());
+        }
+
+        // Write the endoscope program into the SP RAM
+        let start = sys_get_timer().now;
+        if self
+            .swd_bulk_write(endoscope::LOAD, ENDOSCOPE_BYTES)
+            .is_err()
+        {
+            ringbuf_entry!(Trace::InjectionFailed);
+            return Err(());
+        }
+        // log the injection time which can still be improved.
+        let now = sys_get_timer().now;
+        ringbuf_entry!(Trace::Injected {
+            start: endoscope::LOAD,
+            length: ENDOSCOPE_BYTES.len(),
+            delta_t: (now - start) as u32
+        });
+
+        // Resume execution by turning off DHCSR_C_HALT
+        if let Err(e) = self.dp_write_bitflags::<Dhcsr>(Dhcsr::resume()) {
+            ringbuf_entry!(Trace::AckErr(e));
+            return Err(());
+        }
+
+        // It takes about 0.25 seconds (236 RoT systicks) for `endoscope` to run.
+        // Allow about twice that time for the measurement to complete.
+        // endoscope executes a BKPT instruction on completion.
+        // We observe an S_HALT state if all goes well.
+        if self.halt_wait(512).is_err() {
+            ringbuf_entry!(Trace::DidNotHalt);
+            return Err(());
+        };
```

Comment:
What do we observe if endoscope either loops forever or hits a fault?

Reply:
Comments added to document the various cases.